utils/mil_ros_tools/online_bagger/nodes/online_bagger.py

```diff
@@ -0,0 +1,162 @@
+#!/usr/bin/env python
+
+import rospy
+import os
+import rosbag
+import rostopic
+from collections import deque
+import yaml
+import rospkg
+from std_srvs.srv import SetBool
+
+'''
+To call this service call the '/online_bagger' service with the SetBool Parameter set to True.
+Doing this will generate a rosbag with the last x number of seconds saved
+
+for example:
+
+    bagging_server = rospy.ServiceProxy('/online_bagger', SetBool)
+    bag_status = bagging_server(True)
+'''
+
+class OnlineBagger(object):
+    def __init__(self):
+
+        '''
+        - makes dictionary of dequeues
+        - subscribes to set of topics defined by the yaml file in directory
+        - have a streaming option that can be turned on and off in the callback
+        - option to bag data
+        - add a re subscribing option for anything that failed? maybe...?
+        '''
+ 
+        self.failed_topics = []
+        self.n = 0  # number of iterations
+        self.streaming = True
+        self.dumping = False
+        # self.resubscribe
+
+        self.get_params()
+        self.make_dicts()
+        self.subscribe()
+
+
+        self.bagging_service = rospy.Service('/online_bagger', SetBool, self.start_bagging)
+
+    def get_params(self):
+
+        self.topics = rospy.get_param('/online_bagger/message_topics')
+        self.ram_limit = rospy.get_param('/online_bagger/ram_limit')
+        self.stream_time = rospy.get_param('/online_bagger/stream_time')
+
+        rospy.loginfo('stream_time: %s', self.stream_time)
+        # rospy.loginfo('ram_limit: %s', self.ram_limit)
+
+    def make_dicts(self):
+        '''
+        make dictionaries with deques() that will be filled with topics
+        with all of the 5topics defined by the yaml
+        '''
+
+        self.topic_list = {}
+ 
+        for topic in self.topics:
+            self.topic_list[topic] = deque()
+
+        rospy.loginfo('topics: %s', self.topic_list.keys())
+    def subscribe(self):
+        '''
+        Immediately initiated with instance of the BagStream class:
+        subscribes to the set of topics defined in the yaml configuration file
+        '''
+        # use failed topics list that can be resubscribed to at a later time?
+
+        for topic in self.topic_list.keys():
```

Comment:
This will be important for robustness to act on these failed topics. The moment you create the subscriber objects, topics will asynchronously start flowing in, so you could loop here forever until all the topics are subbed to. The only other change would be moving the service provider line before `self.subscribe` was called.